argocd/provider.go

```diff
@@ -269,3 +338,158 @@ func initApiClient(d *schema.ResourceData) (
+				Type:        schema.TypeString,
+				Optional:    true,
+				DefaultFunc: schema.EnvDefaultFunc("KUBE_CLUSTER_CA_CERT_DATA", ""),
+				Description: "PEM-encoded root certificates bundle for TLS authentication.",
+			},
+			"config_paths": {
+				Type:        schema.TypeList,
+				Elem:        &schema.Schema{Type: schema.TypeString},
+				Optional:    true,
+				Description: "A list of paths to kube config files. Can be set with KUBE_CONFIG_PATHS environment variable.",
+			},
+			"config_path": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				DefaultFunc:   schema.EnvDefaultFunc("KUBE_CONFIG_PATH", nil),
+				Description:   "Path to the kube config file. Can be set with KUBE_CONFIG_PATH.",
+				ConflictsWith: []string{"kubernetes.0.config_paths"},
+			},
+			"config_context": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				DefaultFunc: schema.EnvDefaultFunc("KUBE_CTX", ""),
+			},
+			"config_context_auth_info": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				DefaultFunc: schema.EnvDefaultFunc("KUBE_CTX_AUTH_INFO", ""),
+				Description: "",
+			},
+			"config_context_cluster": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				DefaultFunc: schema.EnvDefaultFunc("KUBE_CTX_CLUSTER", ""),
+				Description: "",
+			},
+			"token": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				DefaultFunc: schema.EnvDefaultFunc("KUBE_TOKEN", ""),
+				Description: "Token to authenticate an service account",
+			},
+			"exec": {
+				Type:     schema.TypeList,
+				Optional: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"api_version": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+						"command": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+						"env": {
+							Type:     schema.TypeMap,
+							Optional: true,
+							Elem:     &schema.Schema{Type: schema.TypeString},
+						},
+						"args": {
+							Type:     schema.TypeList,
+							Optional: true,
+							Elem:     &schema.Schema{Type: schema.TypeString},
+						},
+					},
+				},
+				Description: "",
+			},
+		},
+	}
+}
+
+var k8sPrefix = "kubernetes.0."
+
+func k8sGetOk(d *schema.ResourceData, key string) (interface{}, bool) {
+	value, ok := d.GetOk(k8sPrefix + key)
+
+	// For boolean attributes the zero value is Ok
+	switch value.(type) {
+	case bool:
+		// TODO: replace deprecated GetOkExists with SDK v2 equivalent
+		// https://github.com/hashicorp/terraform-plugin-sdk/pull/350
+		value, ok = d.GetOkExists(k8sPrefix + key)
+	}
+
+	// fix: DefaultFunc is not being triggered on TypeList
+	s := kubernetesResource().Schema[key]
+	if !ok && s.DefaultFunc != nil {
+		value, _ = s.DefaultFunc()
+
+		switch v := value.(type) {
+		case string:
+			ok = len(v) != 0
+		case bool:
+			ok = v
+		}
+	}
+
+	return value, ok
+}
+
+func expandStringSlice(s []interface{}) []string {
```

Comment:
Interesting 👍

Reply:
This is copied from the Helm provider, which uses the same logic.